src/dithering_demo/main.cpy

```diff
@@ -0,0 +1,224 @@
+#include "../build/rmkit.h"
+
+#include <algorithm>
+#include <dirent.h>
+#include <string>
+#include <vector>
+
+static const char * IMAGES_DIR = "/opt/etc/dithering_demo"
+
+vector<string> read_directory(const char * dirname):
+  DIR *dir
+  struct dirent *ent
+  vector<string> filenames
+
+  if ((dir = opendir(dirname)) != NULL):
+    while ((ent = readdir (dir)) != NULL):
+      str_d_name := string(ent->d_name)
+      if str_d_name != "." and str_d_name != ".." and ends_with(str_d_name, "png"):
+        filenames.push_back(str_d_name)
+    closedir(dir)
+  else:
+    perror("")
+  sort(filenames.begin(),filenames.end())
+  return filenames
+
+class DitheredBitmap: public ui::Widget:
+  public:
+  image_data image = { NULL }
+  framebuffer::DITHER::MODE dither_mode = framebuffer::DITHER::NONE
+
+  DitheredBitmap(int x, y, w, h): ui::Widget(x, y, w, h):
+    pass
+
+  void render():
+    old_dither_mode := fb->dither
+    fb->dither = self.dither_mode
+    fb->draw_rect(x, y, w, h, WHITE)
+    if self.image.buffer != NULL
+      fb->draw_bitmap(self.image,
+                      self.x + (self.w - self.image.w) / 2,
+                      self.y + (self.h - self.image.h) / 2)
+    fb->dither = old_dither_mode
+
+class IApp:
+  public:
+  virtual void on_image_selected(string) = 0;
+  virtual void on_waveform_selected(int) = 0;
+  virtual void on_dithering_selected(framebuffer::DITHER::MODE) = 0;
+
+class ImageDropdown : public ui::TextDropdown:
+  public:
+  IApp *app
+  vector<string> filenames
+
+  ImageDropdown(int x, y, w, h, IApp *app) \
+      : app(app), \
+      ui::TextDropdown(x, y, w, h, "Image selection"):
+    self.dir = ui::DropdownButton::DIRECTION::DOWN
+    self.populate()
+
+  void populate():
+    section := add_section("Select an image");
+    self.filenames = read_directory(IMAGES_DIR)
+    section->add_options(self.filenames)
+
+  void on_select(int idx):
+    app->on_image_selected(self.filenames[idx])
+
+#define WAVEFORM_PAIR(X) std::make_pair<>( #X, X )
+
+class WaveformDropdown : public ui::TextDropdown:
+  public:
+  IApp *app
+  vector<pair<string,int>> modes = { \
+     WAVEFORM_PAIR(WAVEFORM_MODE_DU), \
+     WAVEFORM_PAIR(WAVEFORM_MODE_GC16), \
+     WAVEFORM_PAIR(WAVEFORM_MODE_GC4), \
+     WAVEFORM_PAIR(WAVEFORM_MODE_A2), \
+     WAVEFORM_PAIR(WAVEFORM_MODE_DU4), \
+     WAVEFORM_PAIR(WAVEFORM_MODE_AUTO) \
+  }
+
+  WaveformDropdown(int x, y, w, h, IApp *app) \
+      : app(app), \
+      ui::TextDropdown(x, y, w, h, "Waveform selection"):
+    self.dir = ui::DropdownButton::DIRECTION::DOWN
+    self.populate()
+
+  void populate():
+    section := add_section("Waveform mode")
+    vector<string> labels
+    for (auto mode : self.modes)
+      labels.push_back(mode.first)
+    section->add_options(labels)
+
+  void on_select(int idx):
+    app->on_waveform_selected(modes[idx].second)
+
+#define DITHER_PAIR(X) std::make_pair<>( #X, framebuffer::DITHER::X )
+
+class DitheringDropdown : public ui::TextDropdown:
```

Comment:
i feel bad about having a whole class for Dropdown instances, i should convert the dropdown to use events too (like button and widget.cpy).

Reply:
that would be awesome! Events definitely make using those widgets a lot easier.

Reply:
d61fa5d37e2f821cfb1bab0617e60b9e622c568b adds selected event to dropdown